4.95.0: compile time warnings

First summary stats:
```console
[tkloczko@devel-g2v SPECS]$ rpmbuild -ba lsof.spec --quiet 2>&1 | grep -- \\[-W | sed 's/.*\[//; s/\]//' | sort | uniq -c | sort -nr
     29 -Wunused-variable
      2 -Wunused-value
      1 -Wparentheses
```
And extracted warnings
```console
util.c:40:13: warning: 'copyright' defined but not used [-Wunused-variable]
   40 | static char copyright[] =
      |             ^~~~~~~~~
In file included from node.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
node.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from store.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
store.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from dstore.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dstore.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from dfile.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dfile.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from dmnt.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dmnt.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from dnode.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dnode.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from usage.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
usage.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from misc.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
misc.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from proc.c:38:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
proc.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
print.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
In file included from main.c:40:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
main.c:35:13: warning: 'copyright' defined but not used [-Wunused-variable]
   35 | static char copyright[] =
      |             ^~~~~~~~~
In file included from dproc.c:37:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dproc.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
arg.c: In function 'ck_file_arg':
arg.c:481:31: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
  481 |             if (!ss & ErrStat == 0)
      |                       ~~~~~~~~^~~~
arg.c: At top level:
arg.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
dsock.c: In function 'process_netsinfo':
dsock.c:1569:24: warning: unused variable 'tp' [-Wunused-variable]
 1569 |         struct tcp_udp*tp;              /* temporary INET socket info pointer */
      |                        ^~
dsock.c: In function 'process_nets6info':
dsock.c:1721:25: warning: unused variable 'tp' [-Wunused-variable]
 1721 |         struct tcp_udp6*tp;             /* temporary INET6 socket info pointer */
      |                         ^~
dsock.c: In function 'get_tcpudp':
dsock.c:78:52: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   78 |                                       + (tp)->laddr,                    \
      |                                                    ^
dsock.c:2999:29: note: in expansion of macro 'TCPUDP_IPC_HASH'
 2999 |                     int i = TCPUDP_IPC_HASH(tp);
      |                             ^~~~~~~~~~~~~~~
dsock.c: In function 'get_tcpudp6':
dsock.c:92:79: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   92 |                                        + TCPUDP6_IPC_ADDR_MK_INT(&(tp)->laddr), \
      |                                                                               ^
dsock.c:3376:29: note: in expansion of macro 'TCPUDP6_IPC_HASH'
 3376 |                     int i = TCPUDP6_IPC_HASH(tp6);
      |                             ^~~~~~~~~~~~~~~~
In file included from dsock.c:38:
lsof.h: At top level:
lsof.h:328:20: warning: 'dummy_utmp' defined but not used [-Wunused-variable]
  328 | static struct utmp dummy_utmp;          /* to get login name length */
      |                    ^~~~~~~~~~
dsock.c:33:13: warning: 'copyright' defined but not used [-Wunused-variable]
   33 | static char copyright[] =
      |             ^~~~~~~~~
```

Thank you. We must set -Wall in our Makefile.

```
arg.c:481:31: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
  481 |             if (!ss & ErrStat == 0)
      |                       ~~~~~~~~^~~~
```
This spots a critical mistake.
I should use `&&`.

It seems that I must release 4.95.1 after fixing it.

```
dsock.c:78:52: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   78 |                                       + (tp)->laddr,                    \
      |                                                    ^
dsock.c:2999:29: note: in expansion of macro 'TCPUDP_IPC_HASH'
```

Incredible mistake.
This mistake makes lsof slower when `[+|-]E` option is given.

--- a/00DIST
+++ b/00DIST
@@ -5336,5 +5336,7 @@ July 14, 2018
 		Inspired by the issue #206 reported by Tomasz Kłoczko (@kloczek).
 
 
+		Show copyright notice in --version output.
+
 The lsof-org team at GitHub
 April 28, 2022

--- a/00DIST
+++ b/00DIST
@@ -5326,5 +5326,15 @@ July 14, 2018
 		comma.
 		Provided by Danny Fowler in #156.
 
+4.9?.?		????? ??, ????
+
+		[linux] fix hash functions used for finding local tcp/udp IPCs
+		There were typos in the code calculating hash values. The typos might
+		break the flatness of hashtables where the endpoint information about
+		locally used tcp/udp was stored. Theoretically, this fix may improve
+		the performance of lsof with [+|-]E option.
+		Inspired by the issue #206 reported by Tomasz Kłoczko (@kloczek).
+
+
 The lsof-org team at GitHub
 April 28, 2022

--- a/00DIST
+++ b/00DIST
@@ -5338,5 +5338,9 @@ July 14, 2018
 
 		Show copyright notice in --version output.
 
+
+		[linux] compile with -Wall option
+
+
 The lsof-org team at GitHub
 April 28, 2022

--- a/arg.c
+++ b/arg.c
@@ -478,7 +478,7 @@ ck_file_arg(i, ac, av, fv, rs, sbp, accept_deleted_file)
 	}
 
 	if (accept_deleted_file) {
-	    if (!ss & ErrStat == 0)
+	    if (!ss && ErrStat == 0)
 		err = 1;
 	    if (ErrStat)
 		ErrStat = 0;

--- a/arg.c
+++ b/arg.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/dialects/linux/Makefile
+++ b/dialects/linux/Makefile
@@ -18,7 +18,7 @@ CDEF=
 CDEFS=  ${CDEF} ${CFGF}
 DEP=	${CFGD} ${CFGDN}
 INCL=	${DINC}
-CFLAGS=	${CDEFS} ${INCL} ${DEP} ${DEBUG}
+CFLAGS=	-Wall ${CDEFS} ${INCL} ${DEP} ${DEBUG}
 
 GRP=
 

--- a/dialects/linux/dfile.c
+++ b/dialects/linux/dfile.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/dialects/linux/dmnt.c
+++ b/dialects/linux/dmnt.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef	lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/dialects/linux/dnode.c
+++ b/dialects/linux/dnode.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 #if	defined(HASEPTOPTS) && defined(HASPTYEPT)

--- a/dialects/linux/dproc.c
+++ b/dialects/linux/dproc.c
@@ -29,11 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
 #include "lsof.h"
 
 

--- a/dialects/linux/dsock.c
+++ b/dialects/linux/dsock.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 #include <sys/xattr.h>
 

--- a/dialects/linux/dsock.c
+++ b/dialects/linux/dsock.c
@@ -1566,7 +1566,6 @@ process_netsinfo(f)
 					 */
 {
 	struct tcp_udp *p;		/* peer INET socket info pointer */
-	struct tcp_udp*tp;		/* temporary INET socket info pointer */
 
 	if (!FeptE)
 	    return;
@@ -1718,7 +1717,6 @@ process_nets6info(f)
 					 */
 {
 	struct tcp_udp6 *p;		/* peer INET6 socket info pointer */
-	struct tcp_udp6*tp;		/* temporary INET6 socket info pointer */
 
 	if (!FeptE)
 	    return;

--- a/dialects/linux/dsock.c
+++ b/dialects/linux/dsock.c
@@ -75,7 +75,7 @@ static char copyright[] =
  * to the same hash backet. This makes seaching the counter part of
  * an end point easier. See get_netpeeri(). */
 #define TCPUDP_IPC_HASH(tp) ((int)(((((tp)->faddr			\
-				      + (tp)->laddr,			\
+				      + (tp)->laddr			\
 				      + (tp)->fport			\
 				      + (tp)->lport			\
 				      + (tp)->proto) * 31415) >> 3)	\
@@ -89,7 +89,7 @@ static char copyright[] =
 	+(int)TCPUDP6_IPC_ADDR_INT32(a, 0x3))
 
 #define TCPUDP6_IPC_HASH(tp) ((int)((((TCPUDP6_IPC_ADDR_MK_INT(&(tp)->faddr) \
-				       + TCPUDP6_IPC_ADDR_MK_INT(&(tp)->laddr), \
+				       + TCPUDP6_IPC_ADDR_MK_INT(&(tp)->laddr) \
 				       + (tp)->fport			\
 				       + (tp)->lport			\
 				       + (tp)->proto) * 31415) >> 3)	\

--- a/dialects/linux/dstore.c
+++ b/dialects/linux/dstore.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1997 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 int HasNFS = 0;				/* NFS mount point status:

--- a/lsof.h
+++ b/lsof.h
@@ -321,12 +321,10 @@ extern int optind;
 
 # if	!defined(LOGINML)
 #  if	defined(HASUTMPX)
-static struct utmpx dummy_utmp;		/* to get login name length */
-#define	LOGINML		sizeof(dummy_utmp.ut_user)
+#define	LOGINML		sizeof(((struct utmpx *)0)->ut_user)
 					/* login name length */
 #  else	/* !defined(HASUTMPX) */
-static struct utmp dummy_utmp;		/* to get login name length */
-#define	LOGINML		sizeof(dummy_utmp.ut_name)
+#define	LOGINML		sizeof(((struct utmp *)0)->ut_name)
 					/* login name length */
 #  endif	/* defined(HASUTMPX) */
 # endif	/* !defined(LOGINML) */

--- a/main.c
+++ b/main.c
@@ -31,12 +31,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/misc.c
+++ b/misc.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 #if	defined(HASWIDECHAR)

--- a/node.c
+++ b/node.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/print.c
+++ b/print.c
@@ -1232,6 +1232,15 @@ addr_too_long:
 				src = 0;
 			}
 		    }
+#else
+		    /*
+		     * Just for suppressing warnings reported from compiler.
+		     *
+		     * src is referenced in lkup_port() only if "!defined(HASNORPC_H)"
+		     * is true. The condition here is !defined(HASNORPC_H) is false.
+		     * Therefore the value of src has no impact.
+		     */
+		    src = 1;
 #endif	/* !defined(HASNORPC_H) */
 
 		    if (strcasecmp(Lf->iproto, "TCP") == 0)

--- a/print.c
+++ b/print.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/print.c
+++ b/print.c
@@ -154,6 +154,7 @@ endnm(sz)
 static void
 fill_portmap()
 {
+	static int already_run = 0;
 	char buf[128], *cp, *nm;
 	CLIENT *c;
 	int h, port, pr;
@@ -168,6 +169,13 @@ fill_portmap()
 	struct sockaddr_in ia;
 	int s = RPC_ANYSOCK;
 #endif	/* !defined(CAN_USE_CLNT_CREATE) */
+
+/*
+ * Make sure this is only run once.
+ */
+	if (already_run)
+		return;
+	already_run = 1;
 
 /*
  * Construct structures for communicating with the portmapper.
@@ -469,7 +477,6 @@ lkup_port(p, pr, src)
 	MALLOC_S nl;
 	char *nm, *pn;
 	static char pb[128];
-	static int pm = 0;
 	struct porttab *pt;
 /*
  * If the hash buckets haven't been allocated, do so.
@@ -502,10 +509,8 @@ lkup_port(p, pr, src)
  * If we're looking up program names for portmapped ports, make sure the
  * portmap table has been loaded.
  */
-	if (FportMap && !pm) {
+	if (FportMap)
 	    (void) fill_portmap();
-	    pm++;
-	}
 #endif	/* !defined(HASNORPC_H) */
 
 /*

--- a/proc.c
+++ b/proc.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 #if	defined(HASEPTOPTS)

--- a/store.c
+++ b/store.c
@@ -29,12 +29,6 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 1994 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
-
 #include "lsof.h"
 
 

--- a/usage.c
+++ b/usage.c
@@ -29,11 +29,8 @@
  * 4. This notice may not be removed or altered.
  */
 
-#ifndef lint
 static char copyright[] =
-"@(#) Copyright 1998 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
+"Copyright 1998 Purdue Research Foundation. All rights reserved.";
 
 #include "lsof.h"
 #include "version.h"
@@ -919,6 +916,7 @@ usage(err, fh, version)
 	 */
 	    (void) fprintf(stderr, "%s version information:\n", Pn);
 	    (void) fprintf(stderr, "    revision: %s\n", LSOF_VERSION);
+	    (void) fprintf(stderr, "    copyright notice: %s\n", copyright);
 	    (void) fprintf(stderr, "    latest revision: %s\n", LSOF_REPO_URL);
 	    (void) fprintf(stderr, "    latest FAQ: %s\n",
 		LSOF_FAQ_URL);

--- a/util.c
+++ b/util.c
@@ -36,11 +36,6 @@
  */
 
 
-#ifndef lint
-static char copyright[] =
-"@(#) Copyright 2008 Purdue Research Foundation.\nAll rights reserved.\n";
-#endif
-
 #if	defined(HAS_STRFTIME)
 #include <time.h>
 #endif	/* defined(HAS_STRFTIME) */

